core/src/main/java/com/datastax/oss/driver/internal/core/cql/DefaultPagingState.java

```diff
@@ -0,0 +1,160 @@
+import com.datastax.oss.driver.api.core.detach.AttachmentPoint;
+import com.datastax.oss.driver.api.core.session.Session;
+import com.datastax.oss.driver.internal.core.data.ValuesHelper;
+import com.datastax.oss.protocol.internal.util.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+
+public class DefaultPagingState implements PagingState {
+
+  private final ByteBuffer rawPagingState;
+  private final byte[] hash;
+  private final int protocolVersion;
+
+  public DefaultPagingState(
+      ByteBuffer rawPagingState, Statement<?> statement, AttachmentPoint attachmentPoint) {
+    this(
+        rawPagingState,
+        hash(statement, rawPagingState, attachmentPoint),
+        attachmentPoint.getProtocolVersion().getCode());
+  }
+
+  private DefaultPagingState(ByteBuffer rawPagingState, byte[] hash, int protocolVersion) {
+    this.rawPagingState = rawPagingState;
+    this.hash = hash;
+    this.protocolVersion = protocolVersion;
+  }
+
+  // Same serialized form as in driver 3:
+  //     size of raw state|size of hash|raw state|hash|protocol version
+  //
+  // The protocol version might be absent, in which case it defaults to V2 (this is for backward
+  // compatibility with 2.0.10 where it is always absent).
+  public static DefaultPagingState fromBytes(byte[] bytes) {
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    short rawPagingStateLength = buffer.getShort();
+    short hashLength = buffer.getShort();
+    int length = rawPagingStateLength + hashLength + 2;
+    int legacyLength = rawPagingStateLength + hashLength; // without protocol version
+    if (buffer.remaining() != length && buffer.remaining() != legacyLength) {
+      throw new IllegalArgumentException(
+          "Cannot deserialize paging state, invalid format. The serialized form was corrupted, "
+              + "or not initially generated from a PagingState object.");
+    }
+    byte[] rawPagingState = new byte[rawPagingStateLength];
+    buffer.get(rawPagingState);
+    byte[] hash = new byte[hashLength];
+    buffer.get(hash);
+    int protocolVersion = buffer.hasRemaining() ? buffer.getShort() : 2;
+    return new DefaultPagingState(ByteBuffer.wrap(rawPagingState), hash, protocolVersion);
+  }
+
+  @Override
+  public byte[] toBytes() {
+    ByteBuffer buffer = ByteBuffer.allocate(rawPagingState.remaining() + hash.length + 6);
+    buffer.putShort((short) rawPagingState.remaining());
+    buffer.putShort((short) hash.length);
+    buffer.put(rawPagingState.duplicate());
+    buffer.put(hash);
+    buffer.putShort((short) protocolVersion);
+    buffer.rewind();
+    return buffer.array();
+  }
+
+  public static DefaultPagingState fromString(String string) {
+    byte[] bytes = Bytes.getArray(Bytes.fromHexString("0x" + string));
```

Comment:
Why do we need to remove `0x` from the output?
I see that this is removed at the reading and writing side (`fromString` and `toString`)
So maybe it's better to keep it?

Reply:
Short answer is we're keeping the format identical to driver 3 :) But I'd say the motivation then was to not include unnecessary information in the string form, it's supposed to be used in URLs and such, and it's already pretty long.